client/alloc_endpoint_test.go

```diff
@@ -446,3 +459,720 @@ func TestAllocations_Stats_ACL(t *testing.T) {
 		require.True(nstructs.IsErrUnknownAllocation(err))
 	}
 }
+
+func TestAlloc_ExecStreaming(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+
+	// Start a server and client
+	s := nomad.TestServer(t, nil)
+	defer s.Shutdown()
+	testutil.WaitForLeader(t, s.RPC)
+
+	c, cleanup := TestClient(t, func(c *config.Config) {
+		c.Servers = []string{s.GetConfig().RPCAddr.String()}
+	})
+	defer cleanup()
+
+	expectedStdout := "Hello from the other side\n"
+	expectedStderr := "Hello from the other side\n"
+	job := mock.BatchJob()
+	job.TaskGroups[0].Count = 1
+	job.TaskGroups[0].Tasks[0].Config = map[string]interface{}{
+		"run_for": "20s",
+		"exec_command": map[string]interface{}{
+			"run_for":       "1ms",
+			"stdout_string": expectedStdout,
+			"stderr_string": expectedStderr,
+			"exit_code":     3,
+		},
+	}
+
+	// Wait for client to be running job
+	testutil.WaitForRunning(t, s.RPC, job)
+
+	// Get the allocation ID
+	args := nstructs.AllocListRequest{}
+	args.Region = "global"
+	resp := nstructs.AllocListResponse{}
+	require.NoError(s.RPC("Alloc.List", &args, &resp))
+	require.Len(resp.Allocations, 1)
+	allocID := resp.Allocations[0].ID
+
+	// Make the request
+	req := &cstructs.AllocExecRequest{
+		AllocID:      allocID,
+		Task:         job.TaskGroups[0].Tasks[0].Name,
+		Tty:          true,
+		Cmd:          []string{"placeholder command"},
+		QueryOptions: nstructs.QueryOptions{Region: "global"},
+	}
+
+	// Get the handler
+	handler, err := c.StreamingRpcHandler("Allocations.Exec")
+	require.Nil(err)
+
+	// Create a pipe
+	p1, p2 := net.Pipe()
+	defer p1.Close()
+	defer p2.Close()
+
+	errCh := make(chan error)
+	frames := make(chan *drivers.ExecTaskStreamingResponseMsg)
+
+	// Start the handler
+	go handler(p2)
+	go decodeFrames(t, p1, frames, errCh)
+
+	// Send the request
+	encoder := codec.NewEncoder(p1, nstructs.MsgpackHandle)
+	require.Nil(encoder.Encode(req))
+
+	timeout := time.After(3 * time.Second)
+
+	exitCode := -1
+	receivedStdout := ""
+	receivedStderr := ""
+
+OUTER:
+	for {
+		select {
+		case <-timeout:
+			// time out report
+			require.Equal(expectedStdout, receivedStderr, "didn't receive expected stdout")
+			require.Equal(expectedStderr, receivedStderr, "didn't receive expected stderr")
+			require.Equal(3, exitCode, "failed to get exit code")
```

Comment:
Might want to make these asserts to get as much output as possible before the timeout.

Reply:
yeah - if we get out the output as expected, we'll below in https://github.com/hashicorp/nomad/pull/5632/files/385cfaf2a3e130b70143cfe47687d053ccda7005#diff-510134c570f501a1e38b3ddd203d466eR560 .  The assertions in timeout is to display the partial result and call out expected output is missing even after timeout.